classes/observers/listeners/VersionPublicationDois.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @file classes/observers/listeners/VerionPublicationDois.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class VersionPublicationDois
+ * @ingroup core
+ *
+ * @brief Listener fired when publication is published
+ */
+
+namespace PKP\observers\listeners;
+
+use APP\core\Application;
+use APP\facades\Repo;
+use APP\publication\Publication;
+use Illuminate\Events\Dispatcher;
+use Illuminate\Support\LazyCollection;
+use PKP\context\Context;
+use PKP\context\ContextDAO;
+use PKP\galley\Galley;
+use PKP\observers\events\PublishedEvent;
+
+class VersionPublicationDois
+{
+    public function subscribe(Dispatcher $events): void
+    {
+        $events->listen(
+            PublishedEvent::class,
+            self::class . '@handlePublishedEvent'
+        );
+    }
+
+    /**
+     * Checks if new publication version should carry over existing DOIs or mint new ones and creates new DOIs as necessary
+     *
+     * NB: The existing doiId should be removed from the latest publication when it matches that of the previous publication,
+     * i.e. it is a copy of the previous publication. This prevents a legitimate DOI from being removed,
+     * e.g. if a versioned publication is unpublished then republished.
+     *
+     */
+    public function handlePublishedEvent(PublishedEvent $event): void
+    {
+        $submission = $event->submission;
+        $publication = $event->newPublication;
+
+        /** @var ContextDAO $contextDao */
+        $contextDao = Application::getContextDAO();
+        /** @var Context $context */
+        $context = $contextDao->getById($submission->getData('contextId'));
+
+        $shouldVersionPublication = $context->getData(Context::SETTING_DOI_VERSIONING);
+        $notFirstPublication = $publication->getData('version') != 1;
```

Comment:
I'm not sure if this is certain to be true. I think it is possible to end up with a different version number even if it is the earliest publication that is currently published. It may be possible through publishing/unpublishing things.

You may need to rely on logic like you use below, where you check if a DOI exists that matches a DOI of a previous publication or something.